CalDAV fix search with limit and time range

kesselb opened the pull request:
<!--
  - 🚨 SECURITY INFO
  -
  - Before sending a pull request that fixes a security issue please report it via our HackerOne page (https://hackerone.com/nextcloud) following our security policy (https://nextcloud.com/security/). This allows us to coordinate the fix and release without potentially exposing all Nextcloud servers and users in the meantime.
-->

* Resolves: https://github.com/nextcloud/calendar/issues/5563
* Resolves: https://github.com/nextcloud/calendar/issues/5695
* Resolves: https://github.com/nextcloud/calendar/issues/5902

## Summary

Event recurrences are evaluated at runtime because the database only knows the first and last occurrence.

Given, a user created 8 events with a yearly reoccurrence and two for events tomorrow.
The upcoming event widget asks the CalDAV backend for 7 events within the next 14 days.

If limit 7 is applied to the SQL query, we find the 7 events with a yearly reoccurrence and discard the events after evaluating the reoccurrence rules because they are not due within the next 14 days and end up with an empty result even if there are two events to show.

The workaround for search requests with a limit and time range is asking for more row than requested and retrying if we have not reached the limit.

## TODO

- [x] https://github.com/nextcloud/calendar/issues/5592 (could be fixed as well)
- [x] https://github.com/nextcloud/calendar/issues/5944 (could be fixed as well)
- [x] https://github.com/nextcloud/calendar/issues/5902 (related, likely not fixed, but I understand the problem now)
- [x] CI
- [x] Review

## Test cases

[search_with_limit_and_timerange.txt](https://github.com/nextcloud/server/files/15240776/search_with_limit_and_timerange.txt)

A sample calendar with 7 reoccurring events (in January 2024) and two events happening this month. 
GitHub does now allow ICS attachments, thus don't forget to rename.

On master/main the upcoming events widget should be empty.
On this branch Cake Tasting and Pasta Day should appear in the widget. 






## Checklist

- Code is [properly formatted](https://docs.nextcloud.com/server/latest/developer_manual/digging_deeper/continuous_integration.html#linting)
- [Sign-off message](https://github.com/src-d/guide/blob/master/developer-community/fix-DCO.md) is added to all commits
- [ ] Tests ([unit](https://docs.nextcloud.com/server/latest/developer_manual/app_development/tutorial.html#unit-tests), [integration](https://docs.nextcloud.com/server/latest/developer_manual/app_development/tutorial.html#integration-tests), api and/or acceptance) are included
- [ ] Screenshots before/after for front-end changes
- [ ] Documentation ([manuals](https://github.com/nextcloud/documentation/) or wiki) has been updated or is not required
- [ ] [Backports requested](https://github.com/nextcloud/backportbot/#usage) where applicable (ex: critical bugfixes)

kesselb committed:
fix(caldav): event search with limit and timerange

Event recurrences are evaluated at runtime because the database only knows the first and last occurrence.

Given, a user created 8 events with a yearly reoccurrence and two for events tomorrow.
The upcoming event widget asks the CalDAV backend for 7 events within the next 14 days.

If limit 7 is applied to the SQL query, we find the 7 events with a yearly reoccurrence and discard the events after evaluating the reoccurrence rules because they are not due within the next 14 days and end up with an empty result even if there are two events to show.

The workaround for search requests with a limit and time range is asking for more row than requested and retrying if we have not reached the limit.

Signed-off-by: Daniel Kesselberg <mail@danielkesselberg.de>

kesselb commented on apps/dav/lib/CalDAV/CalDavBackend.php:
```diff
@@ -1946,52 +1966,38 @@ public function search(
+			for ($attempt = $objectsCount = 0; $attempt < 3 && $objectsCount < $limit; $attempt++) {
+				$objectsCount = array_push($calendarObjects, ...$this->searchCalendarObjectsByQuery($outerQuery, $start, $end));
+				$outerQuery->setFirstResult($offset += $maxResults);
 			}
```
Todo: $objectsCount might be larger than $limit.

kesselb replied:
I think to address https://github.com/nextcloud/calendar/issues/5902 we need to sort the calendar objects by the start date.

kesselb commented:
Reminder for myself:

The search query does not have an "order by" command.
If we are still seeing weirdnesses with upcoming events / search result, we should consider adding one.

ChristophWurst commented:
> The search query does not have an "order by" command.
> If we are still seeing weirdnesses with upcoming events / search result, we should consider adding one.

Databases will probably use an index for the query, or traverse the data according to the primary id. So the sorting will be unpredictable across instances but for the same instance it should be stable.

tcitworld commented on apps/dav/lib/CalDAV/CalDavBackend.php:
```diff
@@ -2031,6 +2039,53 @@ public function search(
+				'name' => 'VCALENDAR',
+				'comp-filters' => [
+					[
+						'name' => 'VEVENT',
```
Should we limit to `VEVENT`s ?
https://icalendar.org/CalDAV-Access-RFC-4791/9-9-caldav-time-range-xml-element.html

kesselb replied:
Well spotted, thank you :pray: 

A [user mentioned](https://github.com/nextcloud/calendar/issues/5563#issuecomment-1900129501) that the widget is not showing items from the tasks app anymore.  I assume that could be the reason. 

What do you suggest, extending to VEVENT, VTODO?

kesselb replied:
I took a look at the missing tasks / the VTODO handling and wrote down my findings at https://github.com/nextcloud/server/issues/45333. 
It's more work than I imagined, and therefore I need to discuss it with Christoph.
Happy to have your feedback / pointers on that one ;)

kesselb commented:
Added the test case "testSearchWithLimitAndTimeRangeShouldReturnObjectsInTheSameOrder". 

Cake Tasting (2024-05-09) and Pasta Day (2024-05-14) match the criteria. Pasta Day was added to the calendar first and therefore is the first result returned by the SQL server. Likely, the SQL server(s) sorted the results by the primary id. 

I observed locally that the records are not always sorted by the primary id. Especially if you move around event's a lot, it happens that suddenly the order changes. 

The calendar search api does not document how/if the results are ordered. 

- A is therefore to drop the test case again.
- B is to extend the pull request (add order by to the query, sort the final result by DTSTART) to have more reliable results for the api consumer.

kesselb committed:
feat(caldav): order the calendar objects by start date for search

Sorting the events by the start date leads to more predictable results for the search API consumers.

Signed-off-by: Daniel Kesselberg <mail@danielkesselberg.de>

kesselb commented on apps/dav/lib/CalDAV/CalDavBackend.php:
```diff
@@ -1946,52 +1966,40 @@ public function search(
-			if ($isValid) {
-				$calendarObjects[] = $row;
-			}
+		if ($hasLimit && $hasTimeRange) {
```
The problem to be solved with this pull request exists only when limit and timerange are used together because the evaluation, is the event within the given timerange, cannot be done with sql.

kesselb commented on apps/dav/lib/CalDAV/CalDavBackend.php:
```diff
@@ -1920,14 +1920,34 @@ public function search(
+		$hasLimit = is_int($limit);
+		$hasTimeRange = false;
+
 		if (isset($options['timerange'])) {
```
Good idea, indeed easier to read now.

ChristophWurst reviewed (approved):
Looks good

kesselb commented:
/backport to stable29

kesselb commented:
/backport to stable28

kesselb commented:
/backport to stable27
